Add recursive simpletags + TrackUID + Targets

aulemahal opened the pull request:
In order to better meet the MKV tag specifications, I added three elements:
- **TrackUID** information for each Track
- **Targets** object with the properties listed in the specifications (Type, TypeValue and UIDs)
-  recursive **Simpletags**, tags can contain other sub-tags stored in a list

[MKV Specs here](https://matroska.org/technical/specs/tagging/index.html)

I also have changes to be able to export MKV objets to a XML format similar to the one used by MKVmerge (for example), but I will put it in another pull request.

aulemahal commented:
Oups... I forgot two methods in MasterElement. 

- **getAll** gets the list of all element with the specified name instead of raising a KeyError, used in the new Targets object to get all UIDs of a certain type.
- **getMasterElements** gets the list of MasterElements that are children of the current MasterElements, used for nested **simpletags**.

Diaoul commented:
Great, thanks for the PR. Can you add a few tests as well so we can make sure everything works as expected?

aulemahal commented:
I just added a test for the targets element. Unfortunately, the test files provided in the Test suite are very simple. The tag targets are not written in the mkv files (except for test5.mkv) and there's no nested simpletags. Therefore, I can't test this with the test suite. 

What I could do : add another mkv test file to the package (with a complex tag system, but almost no video or audio content, to keep the file small).  It would be in the enzyme/test/ folder.

aulemahal commented:
So. 
I found a test mkv file in the same sourceforge directory as the Wave 1 test suite. It covers nested simpletags and different non-default targets. I added the corresponding tests.

I added a UID property to the chapter object. It can be used to have tags specific to a chapter, by using the targets' ChaptersUID list.

Also, put a default value to the targettypevalue, as used in the Mastroka specs and examples.

I think this part is finished. I plan to create another pull request with changes implementing :
- exporting to XML
- attachements objects

If you have any comments and suggestion about my programming style or anything else, feel free!

Voilà

Diaoul commented on enzyme/mkv.py:
```diff
@@ -273,22 +276,51 @@ def fromelement(cls, element):
 
+class Targets(object):
+    """Object for the Targets EBML element"""
+    def __init__(self, targettypevalue=50, targettype=None, trackUIDs=[], chapterUIDs=[], attachementUIDs=[], editionUIDs=[]):
```
Don't use empty lists as default. It won't work as you expect. See http://effbot.org/zone/default-values.htm

Diaoul commented on enzyme/mkv.py:
```diff
@@ -273,22 +276,51 @@ def fromelement(cls, element):
 class SimpleTag(object):
     """Object for the SimpleTag EBML element"""
-    def __init__(self, name, language='und', default=True, string=None, binary=None):
+    def __init__(self, name, language='und', default=True, string=None, binary=None, simpletags=[]):
```
Same as above

Diaoul commented on enzyme/tests/test_mkv.py:
```diff
@@ -14,10 +14,14 @@
             f.extractall(TEST_DIR)
-
+        # Matroska cover_art test file 
+        r2 = requests.get('http://downloads.sourceforge.net/project/matroska/test_files/cover_art.mkv')
```
You can reuse `r`.

Diaoul commented on enzyme/parsers/ebml/core.py:
```diff
@@ -121,6 +121,24 @@ def get(self, name, default=None):
+            raise ValueError('%s elements contain MasterElement' % name)
+        return [e.data for e in elements]
+
+    def getMasterElements(self):
```
Use `lowercase_with_underscore`.

Diaoul commented on enzyme/parsers/ebml/core.py:
```diff
@@ -121,6 +121,24 @@ def get(self, name, default=None):
             raise ValueError('%s is a MasterElement' % name)
         return element.data
 
+    def getAll(self, name, default=[]):
```
Use `lowercase_with_underscore`.

Diaoul commented on enzyme/tests/test_mkv.py:
```diff
@@ -14,10 +14,14 @@
 
 def setUpModule():
     if not os.path.exists(TEST_DIR):
+        # matroska test file Wave 1
```
`# matroska test suite Wave 1`

Diaoul commented on enzyme/tests/test_mkv.py:
```diff
@@ -14,10 +14,14 @@
         with zipfile.ZipFile(io.BytesIO(r.content), 'r') as f:
             f.extractall(TEST_DIR)
-
+        # Matroska cover_art test file 
```
`# matroska cover_art test file`

aulemahal commented:
As I'm not completely use to the Pep8 style I followed the indication of my IDE (SublimePython). Also changed the empty lists to None.

aulemahal commented:
Just began working on my project again after a long pause. I saw that the pull requests wasn't merged. Are there still changes to be made?

I have another pull request ready when this one is merged. I added the capacity to export to XML and to support attachments.

Diaoul commented on enzyme/parsers/ebml/core.py:
```diff
@@ -121,7 +121,7 @@ def get(self, name, default=None):
         return element.data
 
-    def getAll(self, name, default=[]):
+    def get_all(self, name, default=[]):
```
Never use a list as default argument, you can read more on that extensively on the internet

aulemahal commented:
Good. After the weird bug in the Travis CI build, everything is back to normal. And clean and all issues fixed.

aulemahal commented:
@Diaoul I think this pull request is now clean and all issues are fixed. If you want you could merge it.

pannal commented:
@aulemahal @diaoul how hard would it be to support the extraction of certain addon streams, especially SRT/text? 

I'd love that feature for sub zero.

aulemahal commented:
@pannal   I think it wouldn't be that hard for text subtitles. 
Enzyme's EBML parser already has the tools needed to extract the raw track data (which is divided in Clusters/BlockGroups/Blocks in the MKV format). The most challenging part would be to "stitch" the extracted binary blocks into a complete track file. Which is easier if the track is a text file (ex: SRT). 
This said, it would need a lengthy addition to the code and maybe Diaoul's vision of Enzyme does not include such a feature.  For now, I would recommend using subprocess to call mkvextract from python using the track's info from enzyme.

I will look more into it, as I do like the idea of a all-python solution.

pannal commented:
Thanks for your answer. 

I've just found a promising python only library which seems to be able to do just that: https://github.com/QBobWatson/python-ebml. 

It's a full implementation of the EBML specification, as it seems, but I've only had a quick look at it. I'll have to figure out its extent and how it compares to enzyme.

pannal commented:
Also I can't use external tools in the sub zero plugin, because it can't hold binary library dependant data, because it's executed inside the Plex media server python environment and has to be fully portable without non python dependencies.

aulemahal commented:
I made all the same changes and more to another pull request.
